### PR TITLE
Implement conditional permission to use Location permissions in WebViewer

### LINKED ISSUE #2526: Implement conditional permission to use Location permissions in WebViewer
https://github.com/mit-cml/appinventor-sources/blob/dfac595ba77b62f2982f48f41affd625f5b8fa77/appinventor/components/src/com/google/appinventor/components/runtime/WebViewer.java#L479

I think we can make use of conditional permissions here now as implemented in https://github.com/mit-cml/appinventor-sources/pull/1506

### PR DESCRIPTION
**Pull request by devarsh-mavani-19**
In webviewer component the permission for location should be asked only when website requests permission. Fixes #2526 

I have tested in following config
Device: Redmi 6 pro
API level: 28
Android version: Android pie (9)

it would be helpful if someone can test it on other devices as well. Thanks

Note: Haven't updated component & youngandroid versions yet. will do once PR is verified

### CODE REVIEW

**Comment by AppInventorWorkerBee** [comment]
Can one of the admins verify this patch?

**Comment by devarsh-mavani-19** [comment]
@ewpatton I have fixed issue #2654 . can you please verify it? Thanks

**Comment by devarsh-mavani-19** [comment]
@ewpatton can you please review this? Thanks

**Comment by devarsh-mavani-19** [comment]
@ewpatton can you please check this? Thanks

File: appinventor/components/src/com/google/appinventor/components/runtime/WebViewer.java [hunk 1]
```diff
@@ -115,7 +115,9 @@
 // TODO(halabelson): Integrate control of the Back key, when we provide it
 
 @SimpleObject
-@UsesPermissions(permissionNames = "android.permission.INTERNET")
+@UsesPermissions(permissionNames = "android.permission.INTERNET," 
+                  + "android.permission.ACCESS_FINE_LOCATION")
```
Comment (pavi2410) [inline appinventor/components/src/com/google/appinventor/components/runtime/WebViewer.java:119]: Requires `ACCESS_COARSE_LOCATION` permission too

File: appinventor/components/src/com/google/appinventor/components/runtime/WebViewer.java [hunk 2]
```diff
@@ -140,6 +142,11 @@
   // Flag to mark whether we have received permission to read external storage
   private boolean havePermission = false;
 
+  // Flag to check whether app needs location permission
+  private boolean usesLocation = false;
+
+  public ComponentContainer container;
```
Comment (pavi2410) [inline appinventor/components/src/com/google/appinventor/components/runtime/WebViewer.java:148]: Unneccessary

File: appinventor/components/src/com/google/appinventor/components/runtime/WebViewer.java [hunk 3]
```diff
@@ -483,6 +490,18 @@ public void Reload() {
...
+  }
+
+  /**
+   * Determine if the app uses location permissions.
+   * @return true if uses location
+   */
+  @SimpleProperty(userVisible = false,
+      description = "Whether or not to give the application permission to use the Javascript "
+          + "geolocation API. This property is available only in the designer.")
+  public boolean UsesLocation() {
```
Comment (pavi2410) [inline appinventor/components/src/com/google/appinventor/components/runtime/WebViewer.java:503]: Implement conditional permission here.

Ref https://github.com/mit-cml/appinventor-sources/blob/080ec9208b453814fa24f39c6caf53a0eedec421/appinventor/components/src/com/google/appinventor/components/runtime/PhoneCall.java#L161

**Review by pavi2410 — CHANGES_REQUESTED** [review]

File: appinventor/components/src/com/google/appinventor/components/runtime/WebViewer.java (hunk 3 above)
Comment (devarsh-mavani-19) [reply appinventor/components/src/com/google/appinventor/components/runtime/WebViewer.java:503]: @pavi2410 this is a getter method do I need to add conditional permission in getters as well?

Comment (pavi2410) [reply appinventor/components/src/com/google/appinventor/components/runtime/WebViewer.java:503]: No, applying annotation on either of a setter or a getter is fine

Comment (devarsh-mavani-19) [reply appinventor/components/src/com/google/appinventor/components/runtime/WebViewer.java:503]: @pavi2410 ok,  so I just need to add usespermission annotation in either of getter or setter. The conditional requestnig of permission is already handled in EclairUtils file so won't have to add anything else in getter or setter method right? Thanks :-)

Comment (pavi2410) [reply appinventor/components/src/com/google/appinventor/components/runtime/WebViewer.java:503]: Yes, just add the annotation. No more changes required

File: appinventor/components/src/com/google/appinventor/components/runtime/util/EclairUtil.java [hunk 4]
```diff
@@ -62,28 +63,59 @@ public void onGeolocationPermissionsShowPrompt(String origin,
...
-            });
-          alertDialog.setButton(AlertDialog.BUTTON_NEGATIVE, "Refuse",
-            new DialogInterface.OnClickListener() {
-              public void onClick(DialogInterface dialog, int which) {
-                theCallback.invoke(theOrigin, false, true);
-              }
-            });
-          alertDialog.show();
+
+          if (!theCaller.UsesLocation()) {
```
Comment (pavi2410) [inline appinventor/components/src/com/google/appinventor/components/runtime/util/EclairUtil.java:75]: Move this to top (before Line 68)

**Comment by devarsh-mavani-19** [comment]
> @devarsh-mavani-19 What problem is this solving? We keep track of whether or not the WebViewer uses location permission at the project level. If any WebViewer has UsesLocation checked, then we add the permission on Compiler.java on line 377.

Using this PR, we only ask location permission, when requested by webpage (at runtime).

**Comment by jisqyv** [comment]
@devarsh-mavani-19 Before you can request location permission at runtime, you need to also ask for the permission in the Android Manifest, which is compiled into the app. Because location permission is considered "senstive" by Google. We have a property (UsesLocation) that if unchecked results in us *not* asking for permission. This way someone who is using a WebViewer without location doesn't have to argue with Google over the permission. Note: Even with the permission declared in the manifest, the user is still prompted at runtime as well.

**Pull request closed without merging**